### ISSUE TITLE
``fetchSockets()`` always returns ``RemoteSocket[]``

### ISSUE THREAD

**Issue by nathan6am** [socketio/socket.io#4595]
**Describe the bug**
I am trying to retrieve a socket instance by its id, however, using the ``io.fetchSockets()`` method always returns a ``RemoteSocket[]``, so methods like ``socket.timeout`` can't be used from the resulting instance. According to the docs ``fetchSockets`` should return ``Socket[] | RemoteSocket[]``. I am not using an adapter and the behavior is the same in main namespace. 

**To Reproduce**

Please fill the following code example:

Socket.IO server version: `4.5.4`

*Server*

```ts
//Returns a Promise<RemoteSocket | undefined>
const socketInstanceById = async (socketId: string) => {
    const sockets = await io.in(socketId).fetchSockets(); 
    for (const socket of sockets) {
      if (socket.id === socketId) return socket;
    }
 }
```

Reducing the example to 

```ts
const sockets = await io.fetchSockets()
//Type of `sockets` is RemoteSocket[]
```
Is ``fetchSockets`` the only way to retrieve a socket instance by its id? If so is there a way to get a Socket instance and not a RemoteSocket?

**Expected behavior**
``fetchSockets`` should return normal socket instances when called on the same server.

**Comment by darrachequesne** [comment]
Getting a `RemoteSocket` means the client is connected on another server of the cluster.

Regarding the `timeout()` method, it's a known issue that should be fixed soon: https://github.com/socketio/socket.io/pull/4558

**Comment by darrachequesne** [comment]
The type of the `fetchSockets()` response is `RemoteSocket[]`, but it might actually contain some `Socket` instances (for sockets that are on the given server), which share more or less the same interface:

https://github.com/socketio/socket.io/blob/54d5ee05a684371191e207b8089f09fc24eb5107/lib/broadcast-operator.ts#L327-L329

Is there anything that should be exposed by the `RemoteSocket` interface?

**Comment by darrachequesne** [comment]
For future readers:

`remoteSocket.timeout()` was added in https://github.com/socketio/socket.io/commit/0c0eb0016317218c2be3641e706cfaa9bea39a2d, included in version [4.6.0](https://github.com/socketio/socket.io/releases/tag/4.6.0).

Please reopen if needed!

**Comment by TonyLst** [comment]
> Getting a `RemoteSocket` means the client is connected on another server of the cluster.
> 
> Regarding the `timeout()` method, it's a known issue that should be fixed soon: #4558

Is there a way to get complet Socket client instance with "clusterized" fetchSockets() ?

My goal is to access _socket.request.user_ property which is provided by my express-session/passportjs middleware juste as i can access when my app is not clusterized:

`const activeParticipants: Array<Participant> = (
            await browserSocket.adapter.fetchSockets({
                rooms: new Set([code]),
            })
        )
            .map(
                (socketParam) =>
                    (socketParam as unknown as BrowserSocket).request.user
            )
            .filter((user) => !isUserAdmin(user));`

**Comment by darrachequesne** [comment]
Hi! The `request` object is a Node.js `IncomingMessage` message, and cannot be serialized.

You can include the user's id in the `socket.data` attribute:

```js
io.on("connection", (socket) => {
  socket.data.userId = socket.request.user.id;
});
```

And then use it with `fetchSockets()`:

```js
const sockets = await io.fetchSockets();

for (const socket of sockets) {
  console.log(socket.data.userId);
}
```
